pkg/querier/handler/handler.go

```diff
@@ -0,0 +1,604 @@
+	case errorCanceled:
+		code = statusClientClosedConnection
+	case errorTimeout:
+		code = http.StatusServiceUnavailable
+	case errorInternal:
+		code = http.StatusInternalServerError
+	case errorNotFound:
+		code = http.StatusNotFound
+	default:
+		code = http.StatusInternalServerError
+	}
+
+	w.Header().Set("Content-Type", "application/json")
+	w.WriteHeader(code)
+	if n, err := w.Write(b); err != nil {
+		level.Error(api.Logger).Log("msg", "error writing response", "bytesWritten", n, "err", err)
+	}
+}
+
+func parseTimeParam(r *http.Request, paramName string, defaultValue time.Time) (time.Time, error) {
+	val := r.FormValue(paramName)
+	if val == "" {
+		return defaultValue, nil
+	}
+	result, err := parseTime(val)
+	if err != nil {
+		return time.Time{}, errors.Wrapf(err, "Invalid time value for '%s'", paramName)
+	}
+	return result, nil
+}
+
+func parseTime(s string) (time.Time, error) {
+	if t, err := strconv.ParseFloat(s, 64); err == nil {
+		s, ns := math.Modf(t)
+		ns = math.Round(ns*1000) / 1000
+		return time.Unix(int64(s), int64(ns*float64(time.Second))).UTC(), nil
+	}
+	if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
+		return t, nil
+	}
+
+	// Stdlib's time parser can only handle 4 digit years. As a workaround until
+	// that is fixed we want to at least support our own boundary times.
+	// Context: https://github.com/prometheus/client_golang/issues/614
+	// Upstream issue: https://github.com/golang/go/issues/20555
+	switch s {
+	case minTimeFormatted:
+		return minTime, nil
+	case maxTimeFormatted:
+		return maxTime, nil
+	}
+	return time.Time{}, errors.Errorf("cannot parse %q to a valid timestamp", s)
+}
+
+func parseDuration(s string) (time.Duration, error) {
+	if d, err := strconv.ParseFloat(s, 64); err == nil {
+		ts := d * float64(time.Second)
+		if ts > float64(math.MaxInt64) || ts < float64(math.MinInt64) {
+			return 0, errors.Errorf("cannot parse %q to a valid duration. It overflows int64", s)
+		}
+		return time.Duration(ts), nil
+	}
+	if d, err := model.ParseDuration(s); err == nil {
+		return time.Duration(d), nil
+	}
+	return 0, errors.Errorf("cannot parse %q to a valid duration", s)
+}
+
+func createPrometheusResponse(data *queryData) queryrange.PrometheusResponse {
+	if data == nil {
+		return queryrange.PrometheusResponse{
+			Status:    "",
+			Data:      queryrange.PrometheusData{},
+			ErrorType: "",
+			Error:     "",
+			Headers:   []*tripperware.PrometheusResponseHeader{},
+		}
+	}
+
+	sampleStreams := getSampleStreams(data)
+
+	var stats *tripperware.PrometheusResponseStats
+	if data.Stats != nil {
+		builtin := data.Stats.Builtin()
+		stats = &tripperware.PrometheusResponseStats{Samples: getStats(&builtin)}
+	}
+
+	return queryrange.PrometheusResponse{
+		Status: "",
+		Data: queryrange.PrometheusData{
+			ResultType: string(data.ResultType),
+			Result:     *sampleStreams,
+			Stats:      stats,
+		},
+		ErrorType: "",
+		Error:     "",
+		Headers:   []*tripperware.PrometheusResponseHeader{},
```

Comment:
Can we also propagate the headers

Reply:
Headers are added to the grpchttp response (the one that carries the protobuf body) and are later added to PrometheusResponse at the Codec. If I add them to PrometheusResponse directly, should I also add them to the grpchttp response?